### PR TITLE
Add 'How to build a catalog connector' content

### PR DESCRIPTION
**Pull request by ptitzler**
This PR:
 - Adds documentation describing how to build a custom catalog connector
 - Adds quickstart template files
 - Adds basic troubleshooting information
 - Closes https://github.com/elyra-ai/elyra/issues/2249

The entrypoint (for users) is `build-a-custom-connector.md`.

I tested the code by
 - installing the connector from source
 - enabling it
 - creating a KFP pipeline using the included echo component
 - running the pipeline
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

**Comment by ptitzler** [comment]
@kevin-bates fyi

File: component-catalog-connectors/build-a-custom-connector.md [hunk 1]
```diff
@@ -1,3 +1,215 @@
...
+
+   ![server-based catalog connector](doc/images/server-based-catalog-connector.png)
+
+   An example of such a connector is the [Machine Learning Exchange (MLX)](mlx-connector/) connector. This connector uses the MLX API to query the catalog and to retrieve the components that this catalog makes available.
+
+ - Connectors with a built-in catalog: These connectors include a built-in static catalog, eliminating the need for network connectivity once they are installed. This makes them suitable for air gapped environments, but also rather inflexible because a catalog content refresh requires a connector rebuild.
+
+   ![package-based catalog connector](doc/images/package-based-catalog-connector.png)
+
+   The [Apache Airflow component examples connector](airflow-example-components-connector/) and the [Kubeflow Pipelines component examples connector](kfp-example-components-connector/) are two examples. They make a static set of components available.
```
Comment (akchinSTC) [inline component-catalog-connectors/build-a-custom-connector.md:20]:
```suggestion
   The [Apache Airflow component examples connector](airflow-example-components-connector/) and the [Kubeflow Pipelines component examples connector](kfp-example-components-connector/) are two examples. Each connector provides a static set components for each respective runtime.
```

File: component-catalog-connectors/build-a-custom-connector.md [hunk 2]
```diff
@@ -1,3 +1,215 @@
...
+All catalog connectors have in common that they use a schema to define their internal and public properties. The public properties are exposed to the user in the catalog connector configuration, as shown for the quickstart template connector below:
+
+![Quickstart catalog connector configuration](doc/images/quickstart-catalog-configuration.png)
+
+The common public properties are:
+ - "learn more" link, providing access to to connector documentation
+ - catalog instance name
+ - catalog instance description
+ - runtime type filter, which governs the kind of components the connector makes available
+ - component categories, which group the components in the pipeline editor
```
Comment (akchinSTC) [inline component-catalog-connectors/build-a-custom-connector.md:50]:
```suggestion
 - component categories, which group the components together in the pipeline editor
```

File: component-catalog-connectors/build-a-custom-connector.md [hunk 3]
```diff
@@ -1,3 +1,215 @@
...
+
+Connector schemas can optionally also include custom public properties. For illustrative purposes the quickstart template connector defines a required custom property and an optional custom  property. Required properties are typically used to collect catalog connectivity information, such as a server URL, or API token.
```
Comment (akchinSTC) [inline component-catalog-connectors/build-a-custom-connector.md:52]:
```suggestion
Connector schemas can optionally also include custom public properties. For illustrative purposes, the quickstart template connector defines a required custom property and an optional custom property. Required properties are typically used to collect catalog connectivity information, such as a server URL, or API token. 
```

File: component-catalog-connectors/build-a-custom-connector.md [hunk 4]
```diff
@@ -1,3 +1,215 @@
...
+1. [Open the `Catalog components` panel](https://elyra.readthedocs.io/en/latest/user_guide/pipeline-components.html#managing-custom-components-using-the-jupyterlab-ui).
+1. Click `+` and confirm that your connector is displayed in the installed list.
+1. Enable the connector by configuring an instance.
+1. Test the connector.
+
+Refer to the [_Troubleshooting_](#troubleshooting) section for a list of common issues.
+
+### Update the user documentation
+
+A connector should include documentation that enables users to install and uninstall the connector and configure the connector in Elyra.
```
Comment (akchinSTC) [inline component-catalog-connectors/build-a-custom-connector.md:161]:
```suggestion
A connector should include documentation that informs users how to: install and uninstall the connector, configure the connector, and any other important information you wish to relay to the user.
```

File: component-catalog-connectors/connector-template/README.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+This catalog connector enables Elyra to load pipelines components from TODO.
+
+### Install the connector
+
+You can install the TODO catalog connector from PyPI or source code. Note that a **rebuild of JupyterLab is not required**.
+
+**Prerequisites**
+
+- [Elyra](https://elyra.readthedocs.io/en/stable/getting_started/installation.html) (version 3.3 and above).
```
Comment (akchinSTC) [inline component-catalog-connectors/connector-template/README.md:11]:
```suggestion
- [Elyra](https://elyra.readthedocs.io/en/stable/getting_started/installation.html) (version 3.3 and later).
```

File: component-catalog-connectors/build-a-custom-connector.md [hunk 6]
```diff
@@ -1,3 +1,215 @@
...
+ - runtime type filter, which governs the kind of components the connector makes available
+ - component categories, which group the components together in the pipeline editor
+
+Connector schemas can optionally also include custom public properties. For illustrative purposes, the quickstart template connector defines a required custom property and an optional custom property. Required properties are typically used to collect catalog connectivity information, such as a server URL, or API token. 
+
+#### Customize the schema
+
+The quickstart connector default schema file `TODO-catalog.json` is located in the [`todo_catalog_connector`](/component-catalog-connectors/connector-template/todo_catalog_connector/) directory.
+1. Rename the file to reflect your catalog type.
+1. Using an IDE of your choice, find and replace all occurrences of '`TODO-catalog.json`' in all files in the `component-catalog-connectors/connector-template` directory. (There should be seven.)
```
Comment (kiersten-stokes) [inline component-catalog-connectors/build-a-custom-connector.md:58]: I think there are only 3 occurrences of `TODO-catalog.json` in `component-catalog-connectors/connector-template`

File: component-catalog-connectors/build-a-custom-connector.md [hunk 7]
```diff
@@ -1,3 +1,215 @@
...
+1. Open the renamed `TODO-catalog.json` schema file in an editor.
+1. To avoid breaking the connector, do not: 
+     - Remove existing properties, unless indicated otherwise in the comments
+     - Change existing property names, unless indicated otherwise in the comments
+     - Change existing property values, unless indicated otherwise in the comments
+1. Customize the following common catalog connector schema properties:
+   - `$ID`: replace the URL with a valid URL for the renamed schema file.
```
Comment (kiersten-stokes) [inline component-catalog-connectors/build-a-custom-connector.md:65]: NIT: this is lowercase (`$id`) in the schema file

File: component-catalog-connectors/build-a-custom-connector.md [hunk 8]
```diff
@@ -1,3 +1,215 @@
...
+   - `uihints.title`: Replace the placeholder value with the value of the `title` property.
+   - `uihints.reference_url`: This is the "Learn more" link mentioned earlier. Replace the placeholder as desired.
+   - `properties.schemaname.const`: Replace with the new value of `name`.
+   - `properties.metadata.properties.description.default`: This is the default value for the catalog instance description. Replace the placeholder as desired. (Empty value is okay.)  
+   - `properties.metadata.runtime_type.enum`: This list defines the runtime types that the connector supports. The quickstart connector is configured for Kubeflow Pipelines only. Specify `["KUBEFLOW_PIPELINES"]`, `["APACHE_AIRFLOW"]`, or `["KUBEFLOW_PIPELINES", "APACHE_AIRFLOW]`, as appropriate. 
+   - `properties.metadata.runtime_type.default`: The value defines the default selection for `properties.metadata.runtime_type.enum`. (Empty value is okay.)
+   - `properties.metadata.categories.default`: This list defines the categories to which the retrieved components will be assigned by default. The user can override the default selection in the connector's configuration screen.
```
Comment (kiersten-stokes) [inline component-catalog-connectors/build-a-custom-connector.md:74]: Should we include a default in the schema template? Currently there is not one for `categories`. Alternatively, maybe we could just add phrasing to indicate that a user can add a key/value pair for this default field if desired.

File: component-catalog-connectors/build-a-custom-connector.md [hunk 9]
```diff
@@ -1,3 +1,215 @@
...
+   - Remove the properties if your connector does not require any user input to fetch the connector list from the catalog or to retrieve components. This typically only applies to connectors with built-in catalogs, such as the [Kubeflow Pipelines example components connector](/component-catalog-connectors/kfp-example-components-connector/kfp_examples_connector/elyra-kfp-catalog.json).
+   - Otherwise, customize the properties as desired. (You can add as many properties as necessary) You typically need to do this for connectors that access remote catalogs, such as the [MLX connector](/component-catalog-connectors/mlx-connector/mlx_catalog_connector/mlx-catalog.json).
+1. Specify which custom catalog connector schema properties are required. If marked required, the user must provide a non-empty input in the connector configuration.
+   - `properties.metadata.properties.required`: Update this list as appropriate, leaving `"runtime_type"` as is.
```
Comment (kiersten-stokes) [inline component-catalog-connectors/build-a-custom-connector.md:79]: I think this should be:

```suggestion
   - `properties.metadata.required`: Update this list as appropriate, leaving `"runtime_type"` as is.
```

File: component-catalog-connectors/build-a-custom-connector.md [hunk 10]
```diff
@@ -1,3 +1,215 @@
...
+1. Perform a case-insensitive search for `todo` in the file. If one or more matches are still present inthe file verify that you completed all instructions.
```
Comment (kiersten-stokes) [inline component-catalog-connectors/build-a-custom-connector.md:80]:
```suggestion
1. Perform a case-insensitive search for `todo` in the file. If one or more matches are still present in the file verify that you completed all instructions.
```

File: component-catalog-connectors/build-a-custom-connector.md [hunk 11]
```diff
@@ -1,3 +1,215 @@
...
+
+The JupyterLab log file should include an error message that explains what happened. If there is no message, make sure the connector implementation class's `get_catalog_entries` method returns results.
+
+  - Message:
+
+    ```
+    No entrypoint with name 'my-catalog' was found in group 'elyra.component.catalog_types' to match the 'schema_name' given in catalog 'mytestcatalog'. Skipping...
+    ```
+
+    Action:  Verify that `elyra.component.catalog_types` _entrypoint name_ ("my-catalog" in the example above) in `setup.py` matches the value of the `schema_name` property in the connector's schema file.
```
Comment (kiersten-stokes) [inline component-catalog-connectors/build-a-custom-connector.md:203]: Maybe just to make this a little clearer:

```suggestion
    Action: Verify that `elyra.component.catalog_types` _entrypoint name_ ("my-catalog" in the example above) in `setup.py` matches the value of the `properties.schema_name.const` property in the connector's schema file.
```

File: component-catalog-connectors/connector-template/todo_catalog_connector/todo_catalog_connector.py [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+        self.log.debug(f'Value of required parameter \'todo_required_parm\': {todo_required_parm}')
+        self.log.debug(f'Value of optional parameter \'todo_optional_parm\': {todo_optional_parm}')
+
+        # verify that a component key was provided
+        if todo_component_key is None:
+            # Log error and return None
+            self.log.error('Cannot connect to <TODO> catalog: todo_component_key must be provided.')
+            return None
+
+        # TODO: Load component from catalog using the provided key
```
Comment (kiersten-stokes) [inline component-catalog-connectors/connector-template/todo_catalog_connector/todo_catalog_connector.py:111]: To reduce the number of `TODO`s in this file 😄 

```suggestion
        # load component from catalog using the provided key
```

**Review by kiersten-stokes — COMMENTED** [review]
Just a few minor things below. I tried everything out exactly as the instructions indicated and everything is working well!

File: component-catalog-connectors/build-a-custom-connector.md (hunk 7 above)
Comment (kiersten-stokes) [reply component-catalog-connectors/build-a-custom-connector.md:65]:
```suggestion
   - `$id`: replace the URL with a valid URL for the renamed schema file.
```

File: component-catalog-connectors/build-a-custom-connector.md (hunk 6 above)
Comment (ptitzler) [reply component-catalog-connectors/build-a-custom-connector.md:58]: let me change that to several (:

File: component-catalog-connectors/connector-template/todo_catalog_connector/todo_catalog_connector.py (hunk 12 above)
Comment (ptitzler) [reply component-catalog-connectors/connector-template/todo_catalog_connector/todo_catalog_connector.py:111]: Well, it's a todo but fine (:

File: component-catalog-connectors/build-a-custom-connector.md (hunk 8 above)
Comment (ptitzler) [reply component-catalog-connectors/build-a-custom-connector.md:74]: Done! I had an empty default `[]` yesterday, which triggered
```
Validation failed for instance 'hello' using the todo-catalog schema with error: '[]' is not of type 'array'.
```


I've added a value and confirmed that it's working

Comment (kiersten-stokes) [reply component-catalog-connectors/build-a-custom-connector.md:74]: I thought that might have been where that came from 😄

**Review by kiersten-stokes — APPROVED** [review]
LGTM!

**Review by akchinSTC — APPROVED** [review]
👍

**Pull request merged**